pkg/controllermanager/controller/managedseedset/add.go

```diff
@@ -0,0 +1,343 @@
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	seedmanagementv1alpha1 "github.com/gardener/gardener/pkg/apis/seedmanagement/v1alpha1"
+	"github.com/gardener/gardener/pkg/controllerutils/mapper"
+	operationshoot "github.com/gardener/gardener/pkg/operation/shoot"
+	contextutil "github.com/gardener/gardener/pkg/utils/context"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+)
+
+// ControllerName is the name of this controller.
+const ControllerName = "managedseedset"
+
+// AddToManager adds Reconciler to the given manager.
+func (r *Reconciler) AddToManager(mgr manager.Manager) error {
+	if r.Client == nil {
+		r.Client = mgr.GetClient()
+	}
+
+	if r.Actuator == nil {
+		replicaFactory := ReplicaFactoryFunc(NewReplica)
+		replicaGetter := NewReplicaGetter(r.Client, mgr.GetAPIReader(), replicaFactory)
+		r.Actuator = NewActuator(r.Client, replicaGetter, replicaFactory, &r.Config, mgr.GetEventRecorderFor(ControllerName+"-controller"))
+	}
+
+	c, err := builder.
+		ControllerManagedBy(mgr).
+		Named(ControllerName).
+		For(&seedmanagementv1alpha1.ManagedSeedSet{}, builder.WithPredicates(predicate.GenerationChangedPredicate{})).
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: pointer.IntDeref(r.Config.ConcurrentSyncs, 0),
+			RecoverPanic:            true,
+		}).
+		Build(r)
+	if err != nil {
+		return err
+	}
+
+	if err := c.Watch(
+		&source.Kind{Type: &gardencorev1beta1.Shoot{}},
+		&handler.EnqueueRequestForOwner{
+			OwnerType:    &seedmanagementv1alpha1.ManagedSeedSet{},
+			IsController: true,
+		},
+		r.ShootPredicate(),
+	); err != nil {
+		return err
+	}
+
+	if err := c.Watch(
+		&source.Kind{Type: &seedmanagementv1alpha1.ManagedSeed{}},
+		&handler.EnqueueRequestForOwner{
+			OwnerType:    &seedmanagementv1alpha1.ManagedSeedSet{},
+			IsController: true,
+		},
+		r.ManagedSeedPredicate(),
+	); err != nil {
+		return err
+	}
+
+	return c.Watch(
+		&source.Kind{Type: &gardencorev1beta1.Seed{}},
+		mapper.EnqueueRequestsFrom(mapper.MapFunc(r.MapSeedToManagedSeedSet), mapper.UpdateWithNew, c.GetLogger()),
+		r.SeedPredicate(),
+	)
+}
+
+// ShootPredicate returns the predicate for Shoot events.
+// ShootPredicate reacts only on 'CREATE','UPDATE' and 'DELETE' events. It returns true when shoot is deleted.
+// For updates,returns true if:
+// - shoot's health status changed.
+// - shoot belongs to the pending replica and it progressed from the state that caused the replica to be pending.
+func (r *Reconciler) ShootPredicate() predicate.Predicate {
+	return &shootPredicate{}
+}
+
+type shootPredicate struct {
+	ctx    context.Context
+	reader client.Reader
+}
+
+func (p *shootPredicate) InjectStopChannel(stopChan <-chan struct{}) error {
+	p.ctx = contextutil.FromStopChannel(stopChan)
+	return nil
+}
+
+func (p *shootPredicate) InjectClient(client client.Client) error {
+	p.reader = client
+	return nil
+}
+
+func (p *shootPredicate) Create(_ event.CreateEvent) bool { return true }
```

Comment:
Don't we need to perform the checks you currently only do in `Update` also here?